tensorflow/java/src/main/java/org/tensorflow/Tensor.java

```diff
@@ -537,56 +548,72 @@ private static void throwExceptionIfNotByteOfByteArrays(Object array) {
     }
   }
 
+  static HashMap<Class<?>, DataType> class_datatypes = new HashMap<>();
+  static {
+    class_datatypes.put(int.class, DataType.INT32);
+    class_datatypes.put(Integer.class, DataType.INT32);
+    class_datatypes.put(long.class, DataType.INT64);
+    class_datatypes.put(Long.class, DataType.INT64);
+    class_datatypes.put(float.class, DataType.FLOAT);
+    class_datatypes.put(Float.class, DataType.FLOAT);
+    class_datatypes.put(double.class, DataType.DOUBLE);
+    class_datatypes.put(Double.class, DataType.DOUBLE);
+    class_datatypes.put(byte.class, DataType.STRING);
+    class_datatypes.put(Byte.class, DataType.STRING);
+    class_datatypes.put(boolean.class, DataType.BOOL);
+    class_datatypes.put(Boolean.class, DataType.BOOL);
+  }
+
   private static DataType dataTypeOf(Object o) {
-    if (o.getClass().isArray()) {
-      if (Array.getLength(o) == 0) {
-        throw new IllegalArgumentException("cannot create Tensors with a 0 dimension");
-      }
-      // byte[] is a DataType.STRING scalar.
-      Object e = Array.get(o, 0);
-      if (e == null) {
-        throwExceptionIfNotByteOfByteArrays(o);
-        return DataType.STRING;
-      }
-      if (Byte.class.isInstance(e) || byte.class.isInstance(e)) {
-        return DataType.STRING;
-      }
-      return dataTypeOf(e);
+    Class<?> c = o.getClass();
+    while (c.isArray()) {
+      c = c.getComponentType();
     }
-    if (Float.class.isInstance(o) || float.class.isInstance(o)) {
-      return DataType.FLOAT;
-    } else if (Double.class.isInstance(o) || double.class.isInstance(o)) {
-      return DataType.DOUBLE;
-    } else if (Integer.class.isInstance(o) || int.class.isInstance(o)) {
-      return DataType.INT32;
-    } else if (Long.class.isInstance(o) || long.class.isInstance(o)) {
-      return DataType.INT64;
-    } else if (Boolean.class.isInstance(o) || boolean.class.isInstance(o)) {
-      return DataType.BOOL;
-    } else {
-      throw new IllegalArgumentException("cannot create Tensors of " + o.getClass().getName());
+    DataType ret = class_datatypes.get(c);
+    if (ret != null) { 
+      return ret;
     }
+    throw new IllegalArgumentException("cannot create Tensors of type " + c.getName());
   }
 
-  private static int numDimensions(Object o) {
-    if (o.getClass().isArray()) {
-      Object e = Array.get(o, 0);
-      if (e == null) {
-        throwExceptionIfNotByteOfByteArrays(o);
-        return 1;
-      } else if (Byte.class.isInstance(e) || byte.class.isInstance(e)) {
-        return 0;
-      }
-      return 1 + numDimensions(e);
+  /** Returns the number of dimensions of a tensor of type dtype
+   *  when represented by the object o.
```

Comment:
Mind running the [source code formatter](https://github.com/google/google-java-format) on the files?

Reply:
Done.